q2_composition/_ancombc2.py

```diff
@@ -0,0 +1,782 @@
+                    slice_df.rename(
+                        {slice_column: renamed}, axis='columns', inplace=True
+                    )
+
+    return slices
+
+
+def _process_categorical_variables(
+    slices: ANCOMBC2SliceMapping, metadata: qiime2.Metadata
+) -> ANCOMBC2SliceMapping:
+    '''
+    Renames categorical variable columns in each slice in order to make the
+    distinction between a metadata variable and it's given level clear, by
+    separating the two as follows: 'some-variable::some-level'.
+
+    Each column is also annotated with its variable, level, and reference
+    level.
+
+    Parameters
+    ----------
+    slices : ANCOMBC2SliceMapping
+        The raw slices as transformed from the ANCOMBC2 method's output.
+    metadata : qiime2.Metadata
+        The per-sample metadata that contains information about each variable
+        present in the ANCOMBC2 output.
+
+    Returns
+    -------
+    ANCOMBC2SliceMapping
+        The slices with any categorical variable columns renamed and annotated
+        with their reference levels as explained above.
+    '''
+    # restructure categorical variable columns into 'variable::level' format
+    for slice_df in slices.values():
+        for slice_column in slice_df.columns:
+            if _is_categorical(slice_column, metadata):
+                variable, level = _parse_variable_and_level(
+                    slice_column, metadata
+                )
+                slice_df.rename(
+                    {slice_column: variable + '::' + level},
+                    axis='columns',
+                    inplace=True
+                )
+
+    # deduce reference level of each categorical variable and annotate columns
+    # any slice will do except for structural zeros
+    reference_levels = _deduce_reference_levels(slices['lfc'], metadata)
+    for slice_df in slices.values():
+        for column in slice_df.columns:
+            if _is_categorical(column, metadata):
+                variable, level = _parse_variable_and_level(column, metadata)
+                reference_level = reference_levels[variable]
+                slice_df[column].attrs['variable'] = variable
+                slice_df[column].attrs['level'] = level
+                slice_df[column].attrs['reference'] = reference_level
+
+    return slices
+
+
+def _is_categorical(column: str, metadata: qiime2.Metadata) -> bool:
+    '''
+    Determines whether `column` refers to a categorical variable.
+
+    Parameters
+    ----------
+    column : str
+        The column name as returned by ANCOMBC2. If in reference to a
+        categorical column then a level will be appended to the variable name.
+    metadata : qiime2.Metadata
+        The per-sample metadata.
+
+    Returns
+    -------
+    bool
+        Whether `column` refers to a categorical variable.
+    '''
+    # reverse sort to handle md columns that are prefixes of other md columns
```

Comment:
I don't quite understand this use-case - where do we see md cols that are prefixes of other cols?

Reply:
for posterity - @colinvwood and I are discussing this offline, so resolving this as an inline comment